### DISCUSSION TITLE
💡 [v2] Warning: Detected multiple renderers concurrently rendering the same context provider. This is currently unsupported.

### DISCUSSION THREAD

**Discussion by EvanBacon** [expo/router#588] [General]
> This issue is a logic dump

Subsequent static renders of a route are causing the React error/warning:
```
Warning: Detected multiple renderers concurrently rendering the same context provider. This is currently unsupported.
```

The warning log is only ever shown after reloading a statically rendered page that has a React Navigation Tab or Stack layout (i.e. `<Slot />` is fine). I've so far tracked it back to React Navigation's [`getNamedContext`](https://github.com/react-navigation/react-navigation/blob/9fe34b445fcb86e5666f61e144007d7540f014fa/packages/elements/src/getNamedContext.tsx#L10), which is used in `HeaderShownContext` and `HeaderHeightContext`.

I discovered the culprits (after hours of searching), by simply logging them in React. 

```diff
// node_modules/react-dom/cjs/react-dom-server-legacy.browser.development.js
function pushProvider(context, nextValue) {
  var prevValue;

  {
    prevValue = context._currentValue2;
    context._currentValue2 = nextValue;

    {
      if (context._currentRenderer2 !== undefined && context._currentRenderer2 !== null && context._currentRenderer2 !== rendererSigil) {
+        console.log(context, nextValue)
        error('Detected multiple renderers concurrently rendering the ' + 'same context provider. This is currently unsupported.');
      }

      context._currentRenderer2 = rendererSigil;
    }
  }
```

This yields the following:

```
<ref *1> {
  '$$typeof': Symbol(react.context),
  _currentValue: false,
  _currentValue2: true,
  _threadCount: 0,
  Provider: { '$$typeof': Symbol(react.provider), _context: [Circular *1] },
  Consumer: { '$$typeof': Symbol(react.context), _context: [Circular *1] },
  _defaultValue: null,
  _globalName: null,
  _currentRenderer: null,
  _currentRenderer2: {},
  displayName: 'HeaderShownContext'
} true
```

and
 
```
<ref *1> {
  '$$typeof': Symbol(react.context),
  _currentValue: undefined,
  _currentValue2: 64,
  _threadCount: 0,
  Provider: { '$$typeof': Symbol(react.provider), _context: [Circular *1] },
  Consumer: { '$$typeof': Symbol(react.context), _context: [Circular *1] },
  _defaultValue: null,
  _globalName: null,
  _currentRenderer: null,
  _currentRenderer2: {},
  displayName: 'HeaderHeightContext'
} 64
```

Previous attempts at discovery included enabling `source-map-support` for Expo Router so I could see the stack traces, but this didn't work because React disables `source-map-support` before creating the custom stack trace 🫠

# Edit 1

- The temporary solution is to clear the global contexts before rendering https://github.com/expo/router/pull/589

**Comment by satya164** [comment]
https://github.com/facebook/react/pull/22797
